cg/meta/deliver/delivery_api.py

```diff
@@ -0,0 +1,245 @@
+        case_name: str | None = get_delivery_case_name(case=case, pipeline=pipeline)
+        delivery_base = get_delivery_dir_path(
+            case_name=case_name,
+            sample_name=sample.name,
+            customer_id=case.customer.internal_id,
+            ticket=case.latest_ticket,
+            base_path=self.project_base_path,
+        )
+        if not self.dry_run:
+            LOG.debug(f"Creating project path {delivery_base}")
+            delivery_base.mkdir(parents=True, exist_ok=True)
+        return delivery_base
+
+    def _get_deliverable_samples(self, case: Case, pipeline: str) -> Iterable[CaseSample]:
+        return filter(
+            lambda link: self._sample_is_deliverable(link=link, case=case, pipeline=pipeline),
+            self.store.get_case_samples_by_case_id(case.internal_id),
+        )
+
+    def _get_version_for_sample(self, sample: Sample, case: Case, pipeline: str) -> Version:
+        bundle: str = sample.internal_id if pipeline == DataDelivery.FASTQ else case.internal_id
+        return self.hk_api.last_version(bundle)
+
+    def _get_case_files_from_version(
+        self, version: Version, sample_ids: set[str], pipeline: str
+    ) -> Iterable[Path]:
+        """Fetch all case files from a version that are tagged with any of the case tags."""
+
+        if not version:
+            LOG.warning("Version is None, cannot get files")
+            return []
+
+        if not version.files:
+            LOG.warning(f"No files associated with Housekeeper version {version.id}")
+            return []
+
+        version_file: File
+        for version_file in version.files:
+            if not include_file_case(file=version_file, sample_ids=sample_ids, pipeline=pipeline):
+                LOG.debug(f"Skipping file {version_file.path}")
+                continue
+            yield Path(version_file.full_path)
+
+    def _get_sample_files_from_version(
+        self,
+        version: Version,
+        sample_id: str,
+        pipeline: str,
+    ) -> Iterable[Path]:
+        """Fetch files for a sample from a version that are tagged with sample tags."""
+        file_obj: File
+        for file_obj in version.files:
+            if not include_file_sample(file=file_obj, sample_id=sample_id, pipeline=pipeline):
+                continue
+            yield Path(file_obj.full_path)
+
+    def _case_is_deliverable(self, case: Case, pipeline: str) -> bool:
+        last_version = self.hk_api.last_version(case.internal_id)
+        if not last_version:
+            case_tags = get_case_tags_for_pipeline(pipeline)
+            skip_missing = pipeline in constants.SKIP_MISSING or self.ignore_missing_bundles
+            if not case_tags:
+                LOG.info(f"Could not find any version for {case.internal_id}")
+            elif not skip_missing:
+                raise SyntaxError(f"Could not find any version for {case.internal_id}")
+            return False
+        return self._case_has_samples(case)
```

Comment:
There should never be cases with no samples. I went and had a look in the database and any such case found was some sort of manual error. So we can remove `_case_has_samples`. We could then remove one indentation level on the rest. Maybe there are more improvements bu this is already better.
```suggestion
    def _case_is_deliverable(self, case: Case, pipeline: str) -> bool:
        last_version = self.hk_api.last_version(case.internal_id)
        if last_version:
            return True
        case_tags = get_case_tags_for_pipeline(pipeline)
        skip_missing = pipeline in constants.SKIP_MISSING or self.ignore_missing_bundles
        if not case_tags:
            LOG.info(f"Could not find any version for {case.internal_id}")
        elif not skip_missing:
            raise SyntaxError(f"Could not find any version for {case.internal_id}")
        return False
```

Reply:
You and me both, the logic is extremely unclear. I think I would just end up lying if I tried to sort out what this mess is supposed to be doing